include/bm/bm_sim/actions.h

```diff
@@ -749,6 +794,24 @@ class ActionFnEntry {
   ActionData action_data{};
 };
 
+class ActionParamVectorFn : public ActionFn {
+ public:
+  ActionParamVectorFn(ActionFn *action_fn, std::string &primitive_name)
+          : ActionFn(primitive_name, 0, 0), af(action_fn) { }
+
+  void parameter_push_back_action_data(int offset);
+  void parameter_push_back_const(const Data &data);
+  void parameter_push_back_last_header_stack_field(
+      header_stack_id_t header_stack, int field_offset);
+  void parameter_push_back_field(header_id_t h_id, int offset);
+  void parameter_push_back_param_vector();
+
+  size_t get_num_params() const { return af->get_num_params(); }
+
+ private:
+  std::vector<ActionParam> param_vector;
+  ActionFn *af;
```

Comment:
the design doesn't look great to me here: you inherit from a class that you also encapsulate.
The previous version of the code had some duplication, but it made sense to me.